alf/utils/schedulers.py

```diff
@@ -0,0 +1,147 @@
+# limitations under the License.
+"""Schedulers."""
+from alf.trainers.policy_trainer import Trainer
+import gin
+
+
+class Scheduler(object):
+    """Base class of all schedulers.
+
+    A scheduler is used to generate manually defined values based on the training
+    progress.
+
+    The subclass should call ``progress()`` to get the current training progress
+    and use it to calculate the scheduled value. There are three types of training
+    progresses:
+
+    * "percent": percent of training completed.
+    * "iterations": the number training iterations.
+    * "env_steps": the number of environment steps
+
+    """
+
+    def __init__(self, progress_type):
+        """
+        Args:
+            progress_type (str): one of "percent", "iterations", "env_steps"
+        """
+        if progress_type == "percent":
+            self._progress_func = Trainer.progress
+        elif progress_type == "iterations":
+            self._progress_func = Trainer.current_iterations
+        elif progress_type == "env_steps":
+            self._progress_func = Trainer.current_env_steps
+        else:
+            raise ValueError("Unknown progress_type: %s" % progress_type)
+
+    def progress(self):
+        return self._progress_func()
+
+
+class ConstantScheduler(object):
+    def __init__(self, value):
+        self._value = value
+
+    def __call__(self):
+        return self._value
+
+
+@gin.configurable
+class StepScheduler(Scheduler):
+    """There is one value for each defined region of training progress."""
+
+    def __init__(self, progress_type, schedule):
+        """
+        Args:
+            progress_type (str): one of "percent", "iterations", "env_steps"
+            schedule (list[tuple]): each tuple is a pair of (progress, value)
+                which means that if current progress is less than progress,
+                the current value will be value.
+        """
+        super().__init__(progress_type)
+        self._progresses, self._values = zip(*schedule)
+        self._index = 0
+
+    def __call__(self):
+        progress = self.progress()
+        index = self._index
+        progresses = self._progresses
+        while index < len(progresses) - 1 and progress >= progresses[index]:
+            index += 1
+        self._index = index
+        return self._values[index]
+
+
+@gin.configurable
+class LinearScheduler(Scheduler):
+    """The value is linearly changed in each defined region of progress."""
+
+    def __init__(self, progress_type, schedule):
+        """
+        Args:
+            progress_type (str): one of "percent", "iterations", "env_steps"
+            schedule (list[tuple]): each tuple is a pair of (progress, value)
+                which means that if the current progress between progress[i-1]
+                and progress[i], a linear interpolation between value[i-1] and
+                value[i] will be used. progress[0] must be 0.
```

Comment:
Maybe also add a sentence about if the current progress is greater than progress[-1]? (then it's always the same last value)